### PR TITLE
Update api url-standard and urlpattern to use new jsg::Url* impls

### PR DESCRIPTION
**Pull request by jasnell**
Makes use of the new `jsg::Url`, `jsg::UrlSearchParams`, and `jsg::UrlPattern`

This is good for review now but I'm not going to land it right away. I want to go through one final pass on tests to make sure there are no broken edge cases.

And in case you look at the `+17,202` lines changed here, don't worry, most of that is ported tests :-)

### CODE REVIEW

File: src/workerd/jsg/BUILD.bazel [hunk 1]
```diff
@@ -41,8 +41,11 @@ wd_cc_library(
     hdrs = ["url.h"],
     visibility = ["//visibility:public"],
     deps = [
+        "@v8//:v8_icu",
+        "@workerd//:icudata-embed",
```
Comment (Warfields) [inline src/workerd/jsg/BUILD.bazel:45]:
```suggestion
       "@workerd-v8//:v8",
```

File: src/workerd/api/sockets.c++ [hunk 2]
```diff
@@ -179,15 +178,18 @@ jsg::Ref<Socket> connectImplNoOutputLock(
     KJ_CASE_ONEOF(str, kj::String) {
       // We need just the hostname part of the address, i.e. we want to strip out the port.
       // We do this using the standard URL parser since it will handle IPv6 for us as well.
-      auto record = JSG_REQUIRE_NONNULL(url::URL::parse(jsg::usv(kj::str("https://", str))),
+      // It also converts all of the various odd forms of IPv4 addresses.
+      // We use a `fake://` protocol to avoid the automatic port normalization that
```
Comment (mikea) [inline src/workerd/api/sockets.c++:182]: I hope there's a test for this? @dom96

File: src/workerd/api/tests/url-test.js [hunk 3]
```diff
@@ -0,0 +1,9760 @@
...
+    ok(url.searchParams.has('a', 'b'));
+    ok(!url.searchParams.has('a', 'c'));
+    url.searchParams.delete('a', 'c');
+    ok(url.searchParams.has('a'));
+  }
+};
+
+export const urlConstructorTests = {
+  test() {
+    const cases = [
```
Comment (mikea) [inline src/workerd/api/tests/url-test.js:416]: nice. I like data oriented tests very much

File: src/workerd/api/url-standard.c++ [hunk 4]
```diff
@@ -27,2093 +27,280 @@
...
 }
 
-jsg::UsvString URL::getProtocol() {
-  jsg::UsvStringBuilder builder(inner.scheme.size() + 1);
-  builder.addAll(inner.scheme);
-  builder.add(':');
-  return builder.finish();
-}
+#define SET(Name, value) inner.set##Name(value)
+#define GET(Name) inner.get##Name();
```
Comment (mikea) [inline src/workerd/api/url-standard.c++:69]: you don't seem to use GET? at least not in getProtocol, getUsername

File: src/workerd/api/url-standard.c++ [hunk 5]
```diff
@@ -27,2093 +27,280 @@
...
-void URL::setProtocol(jsg::UsvString value) {
-  if (value.empty()) return;
-  jsg::UsvStringBuilder builder(value.size() + 1);
-  builder.addAll(value);
-  builder.add(':');
-  KJ_IF_SOME(record, parse(builder.finish(), nullptr, inner, ParseState::SCHEME_START)) {
-    inner = kj::mv(record);
+
+void URL::setHref(kj::String value) {
+  SET(Href, value);
```
Comment (mikea) [inline src/workerd/api/url-standard.c++:73]: I wonder if these are actually worth it. `inner.setHref(value);` is not to terrible either but absolutely explicit and transparent and has no problems with clangd

File: src/workerd/api/url-standard.h [hunk 6]
```diff
@@ -161,146 +117,71 @@ class URLSearchParams: public jsg::Object {
...
-  V(FILE) \
-  V(FILE_SLASH) \
-  V(FILE_HOST) \
-  V(PATH_START) \
-  V(PATH) \
-  V(OPAQUE_PATH) \
-  V(QUERY) \
-  V(FRAGMENT)
-
 // The humble URL object, in all its spec-compliant glory.
```
Comment (mikea) [inline src/workerd/api/url-standard.h:143]: let's update the comment and mention that all work is delegated to jsg:URL?

File: src/workerd/api/urlpattern-test.js [hunk 7]
```diff
@@ -1,518 +0,0 @@
-// Copyright (c) 2023 Cloudflare, Inc.
```
Comment (mikea) [inline src/workerd/api/urlpattern-test.js:1]: is there a reason we don't like these tests anymore? they would give some additional safety too.

File: src/workerd/jsg/url-test.c++ [hunk 8]
```diff
@@ -100,5 +102,1764 @@ KJ_TEST("Search params (2)") {
...
+    .hostname = kj::str("something"),
+    .pathname = kj::str("something"),
+    .hash = kj::str("something"),
+  }, UrlPattern::ProcessInitOptions {
+    .port = "something"_kj,
+    .search = "something"_kj,
+  })) {
+    KJ_CASE_ONEOF(result, UrlPattern::Init) {
+      KJ_ASSERT(result.baseUrl == kj::none);
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.protocol) == "something");
```
Comment (mikea) [inline src/workerd/jsg/url-test.c++:172]: I think these tests would benefit from the same data driven approach you have in javascript tests. no sure how pleasant this would be in c++, but maybe it is doable?

File: src/workerd/jsg/url.c++ [hunk 9]
```diff
@@ -394,4 +401,1853 @@ kj::Maybe<UrlSearchParams::EntryIterator::Entry> UrlSearchParams::EntryIterator:
   };
 }
 
+// ======================================================================================
```
Comment (mikea) [inline src/workerd/jsg/url.c++:404]: was this code moved from somewhere? reviewing it will be hard...

Also, maybe `url-pattern.c++`?

**Review by mikea — COMMENTED** [review]
I did a first pass. Nice work.

Obviously biggest question is: how are we going to release this? Do you have a plan? @jasnell

File: src/workerd/api/url-standard.c++ [hunk 10]
```diff
@@ -27,2093 +27,280 @@
...
 }
 
-jsg::UsvString URL::getHref() { return inner.getHref(); }
-
-void URL::setHref(jsg::UsvString value) {
-  inner = JSG_REQUIRE_NONNULL(parse(value), TypeError, "Invalid URL string.");
-  KJ_IF_SOME(searchParams, maybeSearchParams) {
-    searchParams->reset(toMaybePtr(inner.query));
-  }
+kj::ArrayPtr<const char> URL::getHref() {
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/api/url-standard.c++:64]: Yes... under the covers in ada, the origin is returned as an `ada_owned_string`, which means the caller is responsible for freeing it, whereas the href is returned as an `ada_string` that points into memory owned by the underlying `ada_url` instance.

File: src/workerd/api/urlpattern-test.js (hunk 7 above)
Comment (jasnell) [reply src/workerd/api/urlpattern-test.js:1]: These tests still exist. They were moved into `tests/url-tests` to keep all of the URL-standard related tests together.

File: src/workerd/jsg/url-test.c++ (hunk 8 above)
Comment (jasnell) [reply src/workerd/jsg/url-test.c++:172]: Doable, yes, I'll save that refactor to a separate PR tho.

File: src/workerd/jsg/url.c++ (hunk 9 above)
Comment (jasnell) [reply src/workerd/jsg/url.c++:404]: Yes and no. This is a majorly refactored implementation of the original api/urlpattern.c++ implementation, updated to use kj::String (instead of jsg::UsvString) and jsg::Url instead of the original api/url-standard.c++ impl. It's changed significantly enough that it is safer to treat it as new code as opposed to moved code.

Comment (jasnell) [reply src/workerd/jsg/url.c++:404]: The reason I kept this in `jsg/url.++` and not a separate `jsg/url-pattern.c++` is really just to limit the number of places where we touch ada-url includes directly and keeping all of the URL-related functionality together. If you feel strongly about it, I could separate this out but I'd just generally prefer not to.

File: src/workerd/api/sockets.c++ [hunk 11]
```diff
@@ -179,15 +178,18 @@ jsg::Ref<Socket> connectImplNoOutputLock(
...
-      // parse to a record whose `port` is set to nullptr.
-      auto port = record.port.orDefault(443);
-      isDefaultFetchPort = port == 443 || port == 80;
-      domain = host.toStr();
+      auto host = parsed.getHostname();
+      auto port = parsed.getPort();
+      JSG_REQUIRE(host != ""_kj, TypeError, "Specified address is missing hostname.");
+      JSG_REQUIRE(port != ""_kj, TypeError, "Specified address is missing a port.");
+      isDefaultFetchPort = port == "443"_kj || port == "80"_kj;
+      domain = kj::str(host, ":", port);
```
Comment (dom96) [inline src/workerd/api/sockets.c++:192]: The `domain` shouldn't contain the port

Comment (jasnell) [reply src/workerd/api/sockets.c++]: Reworked this a bit more to ensure that no matter how the address is passed in the host and port and normalized consistently by passing them through the URL parser.

**Comment by dom96** [comment]
General question: why do we need so many changes to use the new jsg::Url* implementations? This is related to ada-url, right? Why isn't it just a drop-in replacement? I would feel far safer if we could push out as minimal a PR as possible that just switches to ada-url and save the refactoring for a separate PR.

File: src/workerd/api/sockets.c++ (hunk 11 above)
Comment (anonrig) [reply src/workerd/api/sockets.c++]: btw hostname includes host and port.

Comment (jasnell) [reply src/workerd/api/sockets.c++]: `host` is hostname and port, isn't it?

```
> const h = new URL('http://example.org:81')
undefined
> h
URL {
  href: 'http://example.org:81/',
  origin: 'http://example.org:81',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'example.org:81',
  hostname: 'example.org',
  port: '81',
  pathname: '/',
  search: '',
  searchParams: URLSearchParams {},
  hash: ''
}
```

Comment (anonrig) [reply src/workerd/api/sockets.c++]: well, this was bound to happen some time. glad it was you.

File: src/workerd/api/sockets.c++ [hunk 12]
```diff
@@ -172,22 +171,26 @@ jsg::Ref<Socket> connectImplNoOutputLock(
...
   kj::String domain;
   bool isDefaultFetchPort = false;
+
   KJ_SWITCH_ONEOF(address) {
     KJ_CASE_ONEOF(str, kj::String) {
       // We need just the hostname part of the address, i.e. we want to strip out the port.
       // We do this using the standard URL parser since it will handle IPv6 for us as well.
-      auto record = JSG_REQUIRE_NONNULL(url::URL::parse(jsg::usv(kj::str("https://", str))),
+      auto input = kj::str("fake://", str);
+      auto record = JSG_REQUIRE_NONNULL(jsg::Url::tryParse(input.asPtr()),
```
Comment (ohodson) [inline src/workerd/api/sockets.c++:186]: The `UrlRecord` type is dropped in this PR. Suggest renaming `s/record/url/` for clarity or using an explicit type instead of `auto`.

File: src/workerd/api/sockets.c++ [hunk 13]
```diff
@@ -209,10 +212,6 @@ jsg::Ref<Socket> connectImplNoOutputLock(
   JSG_REQUIRE(isValidHost(addressStr), TypeError,
       "Specified address is empty string, contains unsupported characters or is too long.");
 
-  auto& ioContext = IoContext::current();
```
Comment (ohodson) [inline src/workerd/api/sockets.c++:212]: This looks like a policy change and, with only passing familiarity with this code, it would seem to be separate from the URL changes that make up the bulk here.

File: src/workerd/api/url-standard.c++ [hunk 14]
```diff
@@ -27,2093 +27,276 @@
...
-    }
-    if (!it) {
-      // Reached the end.
-      break;
-    }
-    start = ++it;
+jsg::UrlSearchParams initFromSearch(kj::Maybe<kj::ArrayPtr<const char>>& maybeQuery) {
+  KJ_IF_SOME(query, maybeQuery) {
+    return JSG_REQUIRE_NONNULL(jsg::UrlSearchParams::tryParse(query), TypeError,
+                               "Invalud URL search params string.");
```
Comment (ohodson) [inline src/workerd/api/url-standard.c++:150]:
```suggestion
                               "Invalid URL search params string.");
```

File: src/workerd/api/urlpattern.c++ [hunk 15]
```diff
@@ -3,1822 +3,47 @@
...
-      JSG_REQUIRE(baseURL == nullptr, TypeError, BASEURL_ERROR);
-      return init(js, processPatternInit(i, ProcessPatternInitType::PATTERN));
-    }
-  }
-  KJ_UNREACHABLE;
+jsg::Ref<URLPattern> create(jsg::Lock& js, jsg::UrlPattern pattern) {
+  bool ignoreCase = pattern.getIgnoreCase();
+  return jsg::alloc<URLPattern>(kj::mv(pattern),
+      compileRegex(js, pattern.getProtocol(), ignoreCase),
+      compileRegex(js, pattern.getUsername(), ignoreCase),
```
Comment (ohodson) [inline src/workerd/api/urlpattern.c++:30]: This is a bit boggling to the eye. `kj::mv` must happen after all the accessor calls that follow (not a complaint, but jarring to observe with the accessors of `pattern` following due to the execution order).

The forwarded arguments go to macro-ized `UrlPattern` constructor that takes _N_-regex arguments. It would be good to use a corresponding macro pattern, using URL_PATTERN_COMPONENTS, so it is trivially obvious the regexes line up.

File: src/workerd/jsg/url-pattern-test-corpus-failures.h [hunk 16]
```diff
@@ -0,0 +1,433 @@
+
```
Comment (ohodson) [inline src/workerd/jsg/url-pattern-test-corpus-failures.h:1]: Please enable url-pattern-test-corpus-{failure,success}.h self-contained with the appropriate includes. It red-inks the editor otherwise and breaks code navigation.

File: src/workerd/jsg/url.c++ [hunk 17]
```diff
@@ -394,4 +401,1853 @@ kj::Maybe<UrlSearchParams::EntryIterator::Entry> UrlSearchParams::EntryIterator:
...
+inline bool isHexDigit(char c) {
+  // Check if `c` is the ASCII code of a hexadecimal digit.
+  return isAsciiDigit(c) || ('a' <= c && c <= 'f') || ('A' <= c && c <= 'F');
+}
+
+inline bool isAscii(char codepoint) { return codepoint >= 0x00 && codepoint <= 0x7f; };
+
+inline bool isForbiddenHostCodepoint(char c) {
+  return c == 0x00 || c == 0x09 /* Tab */ || c == 0x0a /* LF */ || c == 0x0d /* CR */ ||
+         c == ' ' || c == '#' || c == '%' || c == '/' || c == ':' ||
```
Comment (ohodson) [inline src/workerd/jsg/url.c++:424]: https://url.spec.whatwg.org/#forbidden-host-code-point doesn't have `%` in the forbidden host code point definition.

File: src/workerd/api/sockets.c++ (hunk 13 above)
Comment (jasnell) [reply src/workerd/api/sockets.c++:212]: Not so much a policy change as a minor perf improvement by moving the early exit fiddle check before we do any work here.

File: src/workerd/jsg/url.c++ (hunk 17 above)
Comment (jasnell) [reply src/workerd/jsg/url.c++:424]: This check is only used to work around an inconsistency/bug in the URLPattern spec in how it expects to handle hostnames. It is not used in the actual parsing of URLs. Look at the comment for `isValidHostnameInput` on line 438.

File: src/workerd/jsg/url-pattern-test-corpus-failures.h (hunk 16 above)
Comment (jasnell) [reply src/workerd/jsg/url-pattern-test-corpus-failures.h:1]: better yet, I'm just going to restructure these and inline them to eliminate the extra includes.

File: src/workerd/api/urlpattern.c++ (hunk 15 above)
Comment (jasnell) [reply src/workerd/api/urlpattern.c++:30]: heh, that's fair. Reworking!

File: src/workerd/api/sockets.c++ (hunk 13 above)
Comment (jasnell) [reply src/workerd/api/sockets.c++:212]: For this, I think I'd rather keep this in this PR.

**Review by Warfields — CHANGES_REQUESTED** [review]
Almost Done!

File: src/workerd/api/tests/url-test.js
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/api/tests/url-test.js:1]: It is a port of several tests from several internal ew-test files with a large number of changes/improvements. It's best reviewed as a new chunk

File: src/workerd/jsg/url.c++ [hunk 18]
```diff
@@ -394,4 +401,1853 @@ kj::Maybe<UrlSearchParams::EntryIterator::Entry> UrlSearchParams::EntryIterator:
...
+struct RegexAndNameList {
+  kj::String regex;
+  kj::Array<kj::String> names;
+};
+
+UrlPattern::Result<kj::Array<Token>> tokenize(kj::StringPtr input, Token::Policy policy) {
+  auto it = input.begin();
+  size_t pos = 0;
+  kj::Vector<Token> tokenList(input.size() + 1);
+  // Scan the input and advance both it and pos until the given predicate return false.
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/url.c++:844]: Not sure what you mean by what resources are used. We effectively take the input sting and create a token for each character in the sequence, translating it into whatever sequence of tokens is needed. This follows the URLPattern spec very closely, which is to say that it's not the *most* performant way of handling it but it works.

File: src/workerd/jsg/url.c++ [hunk 19]
```diff
@@ -394,4 +401,1853 @@ kj::Maybe<UrlSearchParams::EntryIterator::Entry> UrlSearchParams::EntryIterator:
...
+    return kj::str(result.slice(0, result.size() - 1));
+  }
+  return kj::none;
+}
+
+kj::Maybe<kj::String> canonicalizeUsername(kj::StringPtr username,
+                                           kj::Maybe<kj::StringPtr> = kj::none) {
+  // @see https://wicg.github.io/urlpattern/#canonicalize-a-username
+  if (username.size() == 0) return kj::str();
+  auto url = KJ_ASSERT_NONNULL(Url::tryParse("fake://dummy.test"_kj));
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/url.c++:480]: compile time, likely not. The canonicalization relies on calling the underlying ada-url APIs which are not defined for consteval, etc.

File: src/workerd/jsg/url.h [hunk 20]
```diff
@@ -178,4 +182,121 @@ inline kj::String KJ_STRINGIFY(const UrlSearchParams& searchParams) {
...
+    // The generated JavaScript regular expression for this component.
+    kj::String regex = nullptr;
+
+    // The list of sub-component names extracted for this component.
+    kj::Array<kj::String> names = nullptr;
+  };
+
+  // A structure providing matching patterns for individual components of a URL.
+  struct Init {
+    kj::Maybe<kj::String> protocol;
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/url.h:224]: Perhaps, that can be a separate change tho.

File: src/workerd/api/sockets.c++ [hunk 21]
```diff
@@ -172,22 +171,26 @@ jsg::Ref<Socket> connectImplNoOutputLock(
     jsg::Lock& js, kj::Maybe<jsg::Ref<Fetcher>> fetcher, AnySocketAddress address,
     jsg::Optional<SocketOptions> options) {
 
+  auto& ioContext = IoContext::current();
+  JSG_REQUIRE(!ioContext.isFiddle(), TypeError, "Socket API not supported in web preview mode.");
+
   // Extract the domain/ip we are connecting to from the address.
   kj::String domain;
   bool isDefaultFetchPort = false;
+
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/api/sockets.c++:180]: Yeah, that was intentional, I prefer a little bit of whitespace here, helps with readability.

File: src/workerd/api/urlpattern.c++ [hunk 22]
```diff
@@ -1828,178 +58,206 @@ kj::Maybe<URLPattern::URLPatternComponentResult> execRegex(
...
+        .baseUrl = kj::mv(i.baseURL),
+      };
+
+      jsg::UrlPattern::ProcessInitOptions options = {
+        .mode = jsg::UrlPattern::ProcessInitOptions::Mode::URL
+      };
+
+      KJ_SWITCH_ONEOF(jsg::UrlPattern::processInit(kj::mv(init), kj::mv(options))) {
+        KJ_CASE_ONEOF(err, kj::String) {
+          JSG_FAIL_REQUIRE(TypeError, kj::mv(err));
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/api/urlpattern.c++:237]: Github is not showing me what line that link is for. Can you give me a file name and line number instead? But in general, no, the throws here are intentional and correct.

File: src/workerd/jsg/url.c++ [hunk 23]
```diff
@@ -394,4 +401,1853 @@ kj::Maybe<UrlSearchParams::EntryIterator::Entry> UrlSearchParams::EntryIterator:
...
+}
+
+kj::String stripSuffixFromProtocol(kj::ArrayPtr<const char> data) {
+  if (data.back() == ':') {
+    return kj::str(data.slice(0, data.size() - 1));
+  }
+  return kj::str(data);
+}
+
+kj::String escape(kj::ArrayPtr<const char> str, auto predicate) {
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/url.c++:581]: Probably but given the size of this already let's do that later.

File: src/workerd/api/tests/url-test.js
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/api/tests/url-test.js:1]: Also, fwiw, I wouldn't bother going into too much of a detailed review on this set of tests. As long as they pass we can work on incremental improvements as necessary

File: src/workerd/api/urlpattern.c++ (hunk 22 above)
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/api/urlpattern.c++:237]: Ah right, that comment is actually outdated and can be removed, I just missed it in the pass. It applied to the actual compilation of the underlying patterns only, not the overall function of this method.

[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/api/urlpattern.c++:236]: Removed the outdated comment!

File: src/workerd/jsg/url.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (Warfields) [reply src/workerd/jsg/url.h]: Yeah this is a big enough PR as it is.

File: src/workerd/jsg/url.c++ (hunk 18 above)
[the remark this answers is not in the document]
Comment (Warfields) [reply src/workerd/jsg/url.c++]: Sounds good, this just reminded me of some very deep theory, so I was wondering if there was an interesting read here. Makes sense that it came from the spec.

**Review by Warfields — APPROVED** [review]
Good job on this one!

**Comment by jasnell** [comment]
Ok, so my plan on this is to merge this only when we are ready to actually cut an internal release specifically with this change so that we can more carefully roll it out in isolation. But otherwise, with exception to the fixup commits that I need to squash, this PR shouild be ready to go (modulo any merge conflicts that sneak in between now and time to merge)

**Comment by jasnell** [comment]
Quick update: I think my plan instead on landing this is going to be to gate this change behind a compile-time flag. While I'm confident there's zero actual risk in this change to any current workers running, we'll want to test a bit carefully internally so the compile time flag will give us the ability to land this here and have it on by default in workerd but selectively enabled in our internal system for a careful/incremental rollout. This will mean adding back in the prior implementation for a short period of time but I'll pull that out entirely once we've fully deployed this change in the production environment.

**Comment by anonrig** [comment]
Is there any update on this @jasnell? Anything I can help with?

**Comment by jasnell** [comment]
This is pending some internal scheduling. The PR is ready to go except maybe a rebase and an updated CI run but we're not quite ready to ship the change in production yet. Looking at next month.
